### ISSUE TITLE
[$1000] Web - Chat - Offline avatar is not updated until page refresh

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#18057]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Login to an account which has more than 20 chats
3. Disable the internet connection
4. Navigate to search page
5. Scroll and select a chat with offline default avatar
6. Enable the internet connection

## Expected Result:
Avatar should be updated to default online image

## Actual Result:
Offline avatar gets grey and is not updated until page refresh.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [x] Windows / Chrome
- [ ] MacOS / Desktop

**Version Number:** 1.3.6.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/234732792-7e5240dd-2119-40a7-aae9-42fae89cfc9d.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0152f84f02db9bee1c</li>
        <li>Upwork Job ID: 1673533068774096896</li>
        <li>Last Price Increase: 2023-07-04</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sophiepintoraetz (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by sophiepintoraetz** [comment]
I'm not able to reproduce this - my avatars updated without having to refresh the page (i.e. I just re-enabled the wifi)

https://user-images.githubusercontent.com/42940078/234776736-539e14fe-d410-46c9-a25f-21cb783400cf.mp4

**Comment by lanitochka17** [comment]
Hello. Issue reproducible. Build 1.3.31.2.

https://github.com/Expensify/App/assets/78819774/a63abe2b-ef6a-49b7-a208-198791367f6f

**Comment by sophiepintoraetz** [comment]
In slack, we agreed that the behaviour should be:
>I agree with showing the first chat clicked and disabling the next clicks.

@mananjadhav - would you mind reviewing the proposals? 🙏

**Comment by sophiepintoraetz** [comment]
Okay yup - I can reproduce this:
![2023-06-27_15-22-02 (1)](https://github.com/Expensify/App/assets/42940078/ac0ed624-1ce2-4625-9261-4f311240c22d)

**Comment by crazy-coding** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Offline avatar is not updated until page refresh

### What is the root cause of that problem?
https://github.com/Expensify/App/blob/e5f00f86fb412702e9c212b5a88766a8ef7da611/src/components/Avatar.js#L67
I noticed that the image section is not reloaded when the connection comes back. This is because the `useOnNetworkReconnect` function does not currently have any functionality to reload the image. I suggest adding a new function to the `useOnNetworkReconnect` function that reloads the image.

### What changes do you think we should make in order to solve the problem?
I see that there is already a `useOnNetworkReconnect` function. I think it would be a good idea to add a new function to this hook that reloads the image. This would make it so that the image is reloaded automatically when the connection comes back.

**Comment by crazy-coding** [comment]
Contributor details
Your Expensify account email: devjames214@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~016561959a9e4febff

**Comment by s77rt** [comment]
@crazy-coding Thanks for the proposal. I wasn't able to reproduce the bug but I doubt your RCA being correct. As per the attached videos the avatar seems to reload but "fails?" with a grey image. Can you explain why is that happening? Is that because we are not passing the proper source or the source is not even loaded (no http request made)

**Comment by s77rt** [comment]
@sophiepintoraetz I'm not able to reproduce this bug. Can we get better/easier reproduction steps?

**Comment by crazy-coding** [comment]
@s77rt This issue is unlikely to occur in production. Because it can occur if the link to the avatar image is broken for any reason. So the browser couldn't load the image but cached a gray image as you can see in the video.

For your reference, here are the steps to reproduce this bug.
- Run the application locally without an internet connection.
- The browser can access the application using a local address, but it will not fetch avatar images from the cloud.
- When you reconnect to the internet, the avatar image is changed to a cached gray (empty) image. This is because of a state variable that triggers the condition for using the default avatar or not.

**Comment by sophiepintoraetz** [comment]
@s77rt - can you upload a video of you attempting to replicate the OP steps on staging? It's difficult to know where to improve the steps because I think they're pretty clear?

**Comment by s77rt** [comment]
@crazy-coding I was able to reproduce finally :grin:. I don't think your RCA is correct. On reconnect the image is fetched again but not displayed. I think this may have to do with `background-image` css style not re-rendering the image but not sure.

https://github.com/Expensify/App/assets/16493223/b2982964-2945-459e-9096-84898d779ae8

**Comment by s77rt** [comment]
@sophiepintoraetz While trying to make a video for how it's not working for me, it worked :sweat_smile:. Thanks!

**Comment by crazy-coding** [comment]
@s77rt I investigated all styles for this bug, but I couldn't find any specific problem with the style.
I think react-native-image is using a cached image even if it loads the image later. I found an interesting thing that the `<img> tag` works normally instead of `<RNImage> component` of `react-native`. Actually, I am not sure this helps but I think we can use `react-native-image-cache` or `react-native-cached-image` library for cleaning cached image for previous URI.

**Comment by s77rt** [comment]
@crazy-coding I think we need to investigate this more before jumping into solutions

**Comment by sophiepintoraetz** [comment]
@s77rt - can we give more specific guidance on what needs to be investigated? If we're not sure, I'll need to take this to an external agency due to it being an older issue (and get it resolved quickly)

**Comment by s77rt** [comment]
@sophiepintoraetz Still looking for proposals that explain the root cause of that bug. This issue went external just 2 days ago so I think we can give it more time.

**Comment by arthurmfgtab** [comment]
Hey, @s77rt  @sophiepintoraetz 
I'm Arthur from Callstack - expert contributor group - and I'd like to take a look at this issue

**Comment by pasyukevich** [comment]
Hi, @s77rt @sophiepintoraetz 
I'm Eugene from Callstack - an expert contributor group. 
I will take this issue in my work from Arthur as now he is involved in another issue.

**Comment by crazy-coding** [comment]
Hi, @s77rt @pasyukevich 
I don't know if "Callstack - an expert contributor group" has more priority than an Upwork freelancer like me.

I have investigated the bug thoroughly, including the exact reason and all possible cases. As I suspected, the bug is caused by a cache problem in the React-Native Image component.

React-Native Image uses `background-image` style instead of the `img` tag, which is common in web browsers. This can lead to a cache problem which is a chronic problem, where the wrong image is displayed in offline situations.
 
I have focused on finding a solution that will avoid caching the wrong image in offline situations. I have written code that successfully solves this problem.

It is important to note that this bug could occur not only with avatar images but with all images on the application. So It is possible that we will need to create a module to implement this functionality.

Here's the actual record and result record.
Actual:

https://github.com/Expensify/App/assets/51104065/362bbbe7-c99e-406a-b8b3-3a8a242c043a

Result:

https://github.com/Expensify/App/assets/51104065/ac75e6be-25cc-4c7c-ba93-6abbc4fcd8d1

**Comment by s77rt** [comment]
@crazy-coding As far as I can tell:
1. User is offline
2. Div is mounted, it uses background-image style
3. The image is being fetched
4. Fetching the image failed
5. Div is updated with fallback icon
6. User is online
7. Div is updated and uses background-image style (same as in 2)
8. Browser sees the same url and remembered that the url is broken
9. Browser does not even make a new request
10. No image is loaded
11. The grey image is not a cached image but just a fill color (from background-color)


The simplest solution I guess is to prevent step 2, if we are offline we should not even bother load the image. Can you please update your proposal based on that?

I was doing some experiments on that as well:

https://github.com/Expensify/App/assets/16493223/d0c9e2ca-d6e1-4d05-a52a-5fc7db1fadb0



BTW, this issue will only be assigned to one of the agencies member if no proposal is accepted. Please update your proposal and tag me again.

**Comment by crazy-coding** [comment]
@s77rt 
[Updated](https://github.com/Expensify/App/issues/18057#issuecomment-1609011748)

I have tested your try-on record and have identified a few unexpected issues. For example, when the user is offline, all images are displayed as the default avatar, even if they have local(svg) avatars. If the user loads images online and then goes offline, all of the avatars will change to the default avatar.

So I suggest a trick. please check on the updated proposal.

**Comment by s77rt** [comment]
@crazy-coding 
> If the user loads images online and then goes offline, all of the avatars will change to the default avatar.

I don't think this is the case. If the images are loaded already then the browser will use the cached version when going offline (unless you disable the cache).


https://github.com/Expensify/App/assets/16493223/a9c656d7-bd6a-4ef0-9027-4b5eaf6c08c5

**Comment by crazy-coding** [comment]
@s77rt Yes, 
I updated the solution by checking the offline status when mounting. Can I share solution here?

**Comment by s77rt** [comment]
@crazy-coding Sharing code diff is prohibited. Just state what needs to be done in plain English or use pseudo code if needed. You can also pinpoint what code needs to be changed. Please update your proposal and tag me again.

**Comment by crazy-coding** [comment]
@s77rt 
I will implement a React hook to check the existence of cached images from the URL when offline. This will ensure that users can still see the images even if they are not connected to the internet. 
If a cached image exists, it will be displayed. Otherwise, a patterned image will be displayed.

**Comment by s77rt** [comment]
@crazy-coding I don't see the point of that. It's seems like unnecessary work. Isn't just preventing the images from being fetched while offline enough?

**Comment by crazy-coding** [comment]
@s77rt Yes, you're right.

**Comment by s77rt** [comment]
@crazy-coding Cool. Can you please update your proposal based on that and let's get this going

**Comment by crazy-coding** [comment]
@s77rt 
[Updated Proposal](https://github.com/Expensify/App/issues/18057#issuecomment-1609011748)

**Comment by s77rt** [comment]
@crazy-coding Thanks, that should do it!

:ribbon: :eyes: :ribbon: C+ reviewed

Let's go with @crazy-coding's proposal https://github.com/Expensify/App/issues/18057#issuecomment-1609011748. Extended RCA and proof of concept solution: https://github.com/Expensify/App/issues/18057#issuecomment-1613904690

**Comment by iwiznia** [comment]
Looks good, but not clear how we will trigger the download of the image after the user comes online. Can you explain that a bit please @crazy-coding ?

**Comment by s77rt** [comment]
@iwiznia We have a hook [useOnNetworkReconnect](https://github.com/Expensify/App/blob/4eef8fe76051707963f650693f31ee1418c5b6b7/src/components/Avatar.js#L67) that will cause a re-render.

**Comment by iwiznia** [comment]
Ah nice!

**Comment by crazy-coding** [comment]
@s77rt @iwiznia I just applied on Upwork.

**Comment by s77rt** [comment]
@sophiepintoraetz @iwiznia We had to close the PR raised by @crazy-coding. We worked on some solutions but none worked flawlessly. We think this is a browser bug (apparently only Chrome). Chrome caches results for requests that are made offline.

Do you think we should close this or re-open for more proposals?

**Comment by sophiepintoraetz** [comment]
@s77rt am I understanding that this is because of a bug in Chrome browser, rather than in the Expensify platform, it's hard to come up with a fix for now?

**Comment by s77rt** [comment]
@sophiepintoraetz Yes, that's it.

**Comment by sophiepintoraetz** [comment]
Wow, okay then. I'll wait for @iwiznia to weigh in but I think the only thing we can do is wait until Chrome releases an update that allows us to fix this? (which feels like a can of worms because when that would be, who knows).

**Comment by iwiznia** [comment]
> Chrome caches results for requests that are made offline.

Very odd, did you find a chrome bug report for this? How do you know it is a bug that will be fixed vs expected behavior that will never change?

What are the alternatives?

**Comment by iwiznia** [comment]
Also, I assume you tested other browsers and the behavior is different there? If so, which ones did you test?

**Comment by s77rt** [comment]
@iwiznia 

> did you find a chrome bug report for this?

No

> How do you know it is a bug that will be fixed vs expected behavior that will never change?

We don't know that

> What are the alternatives?

Prevent image from loading when offline, but this would also mean to not load images that are cached. The drawback is cached images will not work once we go offline.

> I assume you tested other browsers and the behavior is different there? If so, which ones did you test?

Tested on Firefox and WebKit

**Comment by iwiznia** [comment]
> Prevent image from loading when offline, but this would also mean to not load images that are cached. The drawback is cached images will not work once we go offline.

Yeah, that's not great. Any others? 

Also, we sure that's chrome's behavior and that's not some quirk of our code? Can you create a small test file showing the broken behavior of chrome please?

**Comment by iwiznia** [comment]
I did a small test with this file and can't reproduce the behavior you mention:
```
<html>
  <body>
    <img id='a' src="https://www.algosolutions.com/wp-content/uploads/2021/05/algo.png">
<button onClick="document.getElementById('a').remove(); document.body.innerHTML = '<img id=\'a\' src=\'https://www.algosolutions.com/wp-content/uploads/2021/05/algo.png\'>'">load</button>
```
Steps:
- Be offline
- Open the page, see image is not shown
- Connect to the internet
- Press the button
- See image loads

**Comment by s77rt** [comment]
The bug only exists for css urls, use `background-image` style instead of `img` tag. I guess the alternative here is to use some RN image library that does not relay on background-image style or make RNW does not relay on it.

**Comment by iwiznia** [comment]
Ohh I see, can you provide a simple file like I did that shows this behavior please? I am not sure I get exactly what you mean

**Comment by s77rt** [comment]
![Screenshot from 2023-07-12 20-13-13](https://github.com/Expensify/App/assets/16493223/4bcdde00-7fc9-456a-b951-2fc43e1f0709)

RNW seems to use the img tag just for accessibility(?) The `img` seems to have opacity set to `0`. 

https://github.com/necolas/react-native-web/blob/b10c66351eae55a8ea798a1217254d7a2658c5b1/packages/react-native-web/src/exports/Image/index.js#L409

**Comment by s77rt** [comment]
I will work on a better reproduction example that is outside E/App.

**Comment by s77rt** [comment]
@iwiznia Here https://codesandbox.io/s/polished-microservice-f99pf8


https://github.com/Expensify/App/assets/16493223/7253f11d-e51e-4a05-a900-4b74a86832c3

**Comment by s77rt** [comment]
@iwiznia Any thoughts here? Should we look for why RNW uses background-image in the first place? It's pretty weird

**Comment by iwiznia** [comment]
Oh sorry... well I think we are not going to add hacks to work around a browser bug and we probably should instead report this as a bug to chrome and see what happens.

**Comment by s77rt** [comment]
Reported https://bugs.chromium.org/p/chromium/issues/detail?id=1466924

**Comment by iwiznia** [comment]
Awesome thanks!

I doubt this will have much activity for a while since we need to wait what chrome says, so not unassigning myself, but if this needs an engineer in the future, @sophiepintoraetz  please unassign and re-assign another engineer since I am going on sabbatical

**Comment by sophiepintoraetz** [comment]
I'll drop this one to weekly in the meantime - @s77rt let me know how this bug report tracks!

**Comment by sophiepintoraetz** [comment]
Still waiting on Chrome!

**Comment by s77rt** [comment]
Still waiting on Chrome. Keeping this as Monthly seems good.

**Comment by s77rt** [comment]
@thesahindia Seems like a dupe.

@dukenv0307 The reported bug on Chrome has been accepted and it's in open state for now. The best solution is to do nothing. This is not.a bug from our end.

**Comment by sophiepintoraetz** [comment]
Thinking on this, do we close the issue? Seeing as it's still with Chrome (are they guaranteed to fix it?), and we can't do anything about it? @s77rt @iwiznia

**Comment by iwiznia** [comment]
We can do that or keep it open, does not bother much 🤷

**Comment by s77rt** [comment]
I think they have a commit to fix the bug https://chromium.googlesource.com/chromium/src/+/c04f52aa42ef09dd9cc5e50b8b3b1809165a6a3b Not sure if this released to the latest version or not yet (seeing the issue still marked as Open I guess not)

**Comment by izarutskaya** [comment]
Issue is reproducible in build v1.4.69-0
Offline avatar gets grey and is not updated until page refresh.


https://github.com/Expensify/App/assets/115492554/4ad09cdb-55af-4368-ba68-73c8445cf073

**Comment by s77rt** [comment]
@sophiepintoraetz or @iwiznia can you please close this one
